### DISCUSSION TITLE
Yubikey Bad Signature

### DISCUSSION THREAD

**Discussion by DaveSH12** [dani-garcia/vaultwarden#5118] [Q&A] [answered]
Here's the English translation of the provided text:

Hello community, I'm experiencing an issue with Vaultwarden and Yubikey. 
We have our own Yubikey authentication server, and my requests are reaching it, but with a BAD_SIGNATURE error. 
We've discovered that in the Yubico Validation Protocol Version 2, a signature of the actual request is included in the H parameter. 
This is required for authentication. Apparently, Vaultwarden is not sending this signature, causing the server to reject the request. 

Has anyone encountered this problem before, or is there perhaps something I need to configure on my end?

 I've set the Yubikey variable in the Docker Compose file and also configured it in the admin interface.
 # Yubikey Settings
      - YUBICO_CLIENT_ID=ABCDEF
      - YUBICO_SECRET_KEY=123456789
      - YUBICO_SERVER=https://ossmfa-qs.testserver.de/ttype/yubikey

I hope anyone here can help me.

Thanks,
Kind regards,
David

**Comment by BlackDex** [comment]
I'm not sure how any self-hosted/alternative yubikey/yubico OTP Server works.
I just checked where the default is going to, which is `api2.yubico.com/wsapi/2.0/verify`

You might need to point to the correct **verify** endpoint for this to work, as you need to provide the full path to the verify URL.

**Reply by DaveSH12** [reply]
What's the possible length of the Yubikey OPT? We are using 44 characters.

**Reply by BlackDex** [reply]
My key also generates a 44 char length key.

**Reply by DaveSH12** [reply]
Maybe a Problem is that we are using privacyIDEA with all Security Features Enabled. When using api2.yubico.com/wsapi/2.0/verify it seems, that no signature and no client is needed. So authenification an yubico is working and on self hosted OPT Server there are problems :-(

**Reply by BlackDex** [reply]
Looking at the code and the documentation of privacyIDEA, that URL endpoints should be ok. And it does look like we call that endpoint in the correct way.

There might be something incorrectly configured with privacyIDEA maybe?
I would suggest to check the logs of privacyIDEA, and maybe enable debug logging?

Also, looking at your configured mocking data, if that represents what you actually filled-in, then i think you might have swapped the `KEY` and `ID`.

It should look something like this.

```ini
YUBICO_SERVER=https://ossmfa-qs.testserver.de/ttype/yubikey
YUBICO_CLIENT_ID=12345
YUBICO_SECRET_KEY=aBCDeFGh+IjKlmnOPQRST1uVwxy=
``

**Reply by DaveSH12** [reply]
Hey, sorry. It's configured in the right way. I just changed the data for my posting.
In mein Config:
 Yubikey Settings
      - YUBICO_CLIENT_ID=yubikey.apiid.lbavita004
      - YUBICO_SECRET_KEY=4PxXXXXXXXXXXXXXX=

Log Level Debug is activated:


["22.10.2024 12:16:10.310"][rustls::server::hs][DEBUG] decided upon suite TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
["22.10.2024 12:16:10.315"][rustls::server::tls12][DEBUG] Session saved
["22.10.2024 12:16:10.315"][request][INFO] PUT /api/two-factor/yubikey
["22.10.2024 12:16:10.581"][reqwest::connect][DEBUG] starting new connection: https://ossmfa-qs.testserver.de/
["22.10.2024 12:16:10.707"][error][ERROR] Invalid Yubikey OTP provided.
[CAUSE] BadSignature
["22.10.2024 12:16:10.707"][response][INFO] (activate_yubikey_put) PUT /api/two-factor/yubikey => 400 Bad Request
["22.10.2024 12:16:10.708"][rustls::common_state][DEBUG] Sending warning alert CloseNotify
["22.10.2024 12:16:11.228"][rustls::server::hs][DEBUG] decided upon suite TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
["22.10.2024 12:16:11.233"][rustls::server::tls12][DEBUG] Session saved
["22.10.2024 12:16:11.234"][request][INFO] GET /icons/lbavita004.testserver.de/icon.png
["22.10.2024 12:16:11.234"][reqwest::connect][DEBUG] starting new connection: https://lbavita004.testserver.de/
["22.10.2024 12:16:11.234"][reqwest::connect][DEBUG] starting new connection: https://lbavita004.testserver.de/
["22.10.2024 12:16:11.234"][vaultwarden::api::icons][WARN] IP 10.62.32.50 for domain 'lbavita004.testserver.de' is not a global IP!
["22.10.2024 12:16:11.234"][response][INFO] (icon_internal) GET /icons/<domain>/icon.png => 200 OK
["22.10.2024 12:16:11.235"][rustls::common_state][DEBUG] Sending warning alert CloseNotify

**Reply by BlackDex** [reply]
Is the `ossmfa-qs.testserver.de` server using a self-signed certificate?

**Reply by DaveSH12** [reply]
We have our own CA in our company and our servers are using the certificates, generated by our CA

**Reply by BlackDex** [reply]
Might be that Vaultwarden doesn't trust this certificate.
What happens if you browse to this `https://vw.your-domain.tld/icons/ossmfa-qs.testserver.de/icon.png`, you already have DEBUG enabled, that might be enough and see if that generates some error.

Or, if running in a container, try to run `curl -v https://ossmfa-qs.testserver.de` within the container and see if that generates a certificate error. If that is the case, then i think Vaultwarden doesn't trust that certificate, and you need to add that CA to the trusted CA's

**Reply by DaveSH12** [reply]
Vaultwarden is running in a Docker Container. I run the curl command within the container.
SSL Certificate seems to be ok.

root@lbavita004:/homes/MGMT/jan-admin# docker exec -it vaultwarden curl -v https://ossmfa-qs.testserver.de
*   Trying 10.62.139.47:443...
* Connected to ossmfa-qs.testserver.de (10.XX.XXX.XX) port 443 (#0)
* ALPN: offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: C=DE; ST=XXXXXXX; L=xxxxxx; O=xxxxxx; CN=ossmfa-qs.testserver.de
*  start date: Feb  8 10:30:22 2024 GMT
*  expire date: May 13 10:30:22 2026 GMT
*  subjectAltName: host "ossmfa-qs.testserver.de" matched cert's "ossmfa-qs.testserver.de"
*  issuer: C=DE; O=XXXX; CN=xxxxx CA 04
*  SSL certificate verify ok.
* using HTTP/1.1
> GET / HTTP/1.1
> Host: ossmfa-qs.testserver.de
> User-Agent: curl/7.88.1
> Accept: */*
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
< HTTP/1.1 200 OK
< Server: nginx
< Date: Tue, 22 Oct 2024 12:37:36 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 9543
< Connection: keep-alive
< Feature-Policy: geolocation 'none'; midi 'none'; camera 'none'; usb 'none'; magnetometer 'none'; accelerometer 'none'; vr 'none'; speaker 'none'; ambient-light-sensor 'none'; gyroscope 'none'; microphone 'none'
< Strict-Transport-Security: max-age=15768000
< X-XSS-Protection: 1; mode=block
< Referrer-Policy: strict-origin-when-cross-origin
< Content-Security-Policy: default-src 'self' http: https: ws: wss: data: blob: 'unsafe-inline' 'unsafe-eval'; frame-ancestors 'self';
< X-Frame-Options: DENY
< X-Content-Type-Options: nosniff
<

**Reply by DaveSH12** [reply]
Hi,
when trying to set OTP via Yubikey (generating 44 signs) in Vaultwarden I always get  "Invalid Yubikey OPT Provided".
Wenn only using the 12 chars from the Public Key, I can save the Key. But login after this is not possible.
I don't know it there is an issue with the 44char Key. It seems that Vaultwarden can not save this. 

Is Vaultwarden Using the parameter H for the optional HMAC-SHA1 signature for the request.

**Comment by DaveSH12** [comment]
Hey BlackDex,
I have now some Information from my colleages, regarding the logfiles from our own Authentication Server:

The Request seems to be ok, in this code line the error occurs:
[privacyidea/privacyidea/lib/tokens/yubikeytoken.py at 3712c767615691e0928bf9698670bdec29555162](https://github.com/privacyidea/privacyidea/blob/3712c767615691e0928bf9698670bdec29555162/privacyidea/lib/tokens/yubikeytoken.py#L397)

This is the validation algoritm:
[https://github.com/privacyidea/privacyidea/blob/3712c767615691e0928bf9698670bdec29555162/privacyidea/lib/tokens/yubikeytoken.py#L79](https://github.com/privacyidea/privacyidea/blob/3712c767615691e0928bf9698670bdec29555162/privacyidea/lib/tokens/yubikeytoken.py#L79) 

Can you please take a look to the class, to verify if the validation/signature has any differents on your site?

**Reply by BlackDex** [reply]
That only thing that i can think of is that the base64 is somehow different which you provide and which is stored in privacyIDEA.
I would double check if everything is 100% exactly the same. Because if the Base64 encoded API_KEY is even one character off, it will break everything.

**Comment by fxwgr** [comment, marked as answer]
@DaveSH12 I stumbled across the same problem which you are facing.
It is because privacyIDEA does not only calculate your given keys, it also appends an own key (ttype), which should not be there imho, as it is not according the Yubico specification.
I also have opened a PR for fixing it: https://github.com/privacyidea/privacyidea/pull/4307
With that fix, everything is working fine and validations are now correct as they should be.

**Reply by DaveSH12** [reply]
@fxwgr Thank you, that was what I am looking for.
